tensorwaves/physics/helicityformalism/amplitude.py

```diff
@@ -1,246 +1,402 @@
+        # otherwise the memory footprint can increase dramatically
+        newdataset = {x: tf.constant(y) for x, y in dataset.items()}
+        return self._model(newdataset).numpy()
 
-    :param theta:
-    :param j:
-    :param m1:
-    :param m2:
+    @property
+    def parameters(self) -> dict:
+        return {x: y.value().numpy() for x, y in self.__parameters.items()}
 
+    def update_parameters(self, new_parameters: dict) -> None:
+        for name, value in new_parameters.items():
+            self.__parameters[name].assign(value)
+
+
+class IntensityBuilder:
+    """Builds Intensities from helicity formalism recipe files.
+
+    Args:
+        particles: A dictionary containing info of various particles.
+        kinematics: A helicity kinematics instance.
+        phsp_data: An optional phase space event collection, which is required
+            if a normalization of the Intensity is performed.
     """
-    from sympy import Rational
-    from sympy.abc import x
-    from sympy.utilities.lambdify import lambdify
-    from sympy.physics.quantum.spin import Rotation as Wigner
-
-    d = (
-        Wigner.d(Rational(j, 2), Rational(m1, 2), Rational(m2, 2), x)
-        .doit()
-        .evalf()
-    )
-    return lambdify(x, d, "tensorflow")(theta)
-
-
-def create_intensity(recipe: dict):
-    if "Intensity" in recipe:
-        recipe = recipe["Intensity"]
-
-    # return _TFNormalizedIntensity(_create_intensity(recipe), None)
-    return _create_intensity(recipe)
-
-
-def _create_intensity(recipe: dict):
-    intensity_class = recipe["Class"]
-    logging.info("creating", intensity_class)
-
-    # this dict could be used for builder injections later on
-    known_intensity_builders = {
-        "IncoherentIntensity": _IncoherentIntensity,
-        "CoherentIntensity": _CoherentIntensity,
-        "StrengthIntensity": _StrengthIntensity,
-        "NormalizedIntensity": _NormalizedIntensity,
-    }
-    if intensity_class in known_intensity_builders:
-        return known_intensity_builders[intensity_class](recipe)
-    else:
-        logging.error(
-            "unknown intensity " + str(intensity_class) + "! Skipping"
+
+    def __init__(
+        self,
+        particles: dict,
+        kinematics: HelicityKinematics,
+        phsp_data: dict = None,
+    ):
+        self._particles = particles
+        self._dynamics: Dict[str, Any] = {}
+        self._kinematics = kinematics
+        self._parameters: Dict[str, tf.Variable] = {}
+        self._phsp_data = phsp_data
+
+    def create_intensity(self, recipe: dict) -> IntensityTF:
+        """Create an `IntensityTF` instance based on a recipe.
+
+        Args:
+            recipe: A dict containing builder instructions. These recipe
+                files can be generated via the expert system
+                (see `~.expertsystem.amplitude.helicitydecay.HelicityAmplitudeGeneratorXML`)
+
+        """
+        self._dynamics = recipe["Dynamics"]
```

Comment:
We may have to create some sort of object representing a recipe structure, because it's hard to debug string keys. In general it is not advised to work with a nested structure of dictionaries for a specific data structure. In addition, it assumes to a large degree we are working with YAML recipes here.

Reply:
> On the top of my head I only see that you have no "tab completion" for the contents if you use strings.

It's also the fact that you will find typos in the strings only at runtime.

> How so? We only use a dict of strings and basic python types.

I remember mainly the mess of using XML, where there would be tags and attributes, some with an `@` prepended.

Reply:
> It's also the fact that you will find typos in the strings only at runtime.

Yeah, thats what I mean with "tab completion" :D. But to test the functionality you also have to write unit tests, right? And the whole idea is that the recipe is dynamic. So I will only know at runtime (once I pass it to the builder) what is actually inside this. Maybe we should move this discussion to a separate issue

>I remember mainly the mess of using XML, where there would be tags and attributes, some with an @ prepended.

Yeah that was xml language specific (attribute). Now I just require a nested dictionary of those special keys. So it should also work fine with the xml, if the attributes are changed into tags.